chat-plugins/scavengers.js

```diff
@@ -1,303 +1,603 @@
-			let uptime = (Date.now() - scavengers.startTime) / 1000;
-			let uptimeText;
-			if (uptime > 24 * 60 * 60) {
-				let uptimeDays = Math.floor(uptime / (24 * 60 * 60));
-				uptimeText = uptimeDays + " " + (uptimeDays === 1 ? "day" : "days");
-				let uptimeHours = Math.floor(uptime / (60 * 60)) - uptimeDays * 24;
-				if (uptimeHours) uptimeText += ", " + uptimeHours + " " + (uptimeHours === 1 ? "hour" : "hours");
-			} else {
-				uptimeText = Chat.toDurationString(uptime * 1000);
+	}
+
+	onSendQuestion(user) {
+		if (!(user.userid in this.players) || user.userid === this.hostId) return false;
+
+		let player = this.players[user.userid];
+		if (player.completed) return false;
+
+		let current = player.getCurrentQuestion();
+
+		player.sendRoom(`You are on ${(current.number === this.questions.length ? "final " : "")}hint #${current.number}: ${current.question.hint}`);
+		return true;
+	}
+
+	onComplete(player) {
+		if (player.completed) return false;
+
+		let now = Date.now();
+		let time = Chat.toDurationString(now - this.startTime, {hhmmss: true});
+
+		let blitz = this.isOfficial && now - this.startTime <= 60000;
+
+		player.completed = true;
+		this.completed.push({name: player.name, time: time, blitz: blitz});
+		let place = formatOrder(this.completed.length);
+
+		this.announce(`<em>${Chat.escapeHTML(player.name)}</em> has finished the hunt in ${place} place! (${time}${(blitz ? " - BLITZ" : "")})`);
+	}
+
+	onEnd(reset, endedBy) {
+		if (!reset) {
+			// give points for winning and blitzes in official games
+			if (this.isOfficial) {
+				for (let i = 0; i < this.completed.length; i++) {
+					if (!this.completed[i].blitz && i > 5) break; // there won't be any more need to keep going
+					let name = this.completed[i].name;
+					if (points[i]) addPoints(name, points[i]);
+					if (this.completed[i].blitz) addPoints(name, 10);
+				}
+				writePoints();
 			}
-			this.sendReply('The current scavenger hunt has been up for: ' + uptimeText);
-			if (scavengers.finished.length) {
-				this.sendReply('The following users have completed it: ' + scavengers.finished.join(', '));
-			} else {
-				this.sendReply('No user has completed it.');
+
+			this.announce(
+				`The scavenger hunt has ended ${(endedBy ? "by " + Chat.escapeHTML(endedBy.name) : "automatically")}.<br />` +
+				`${this.completed.slice(0, 3).map((p, i) => `${formatOrder(i + 1)} place: <em>${Chat.escapeHTML(p.name)}</em>.<br />`).join("")}${this.completed.length > 3 ? `Consolation Prize: ${this.completed.slice(3).map(e => Chat.escapeHTML(e.name)).join(",")}<br />` : ""}<br />` +
+				`<details><summary>Solution: </summary><br />${this.questions.map((q, i) => `${i + 1}) ${Chat.escapeHTML(q.hint)} <span style="color: lightgreen">[<em>${Chat.escapeHTML(q.answer)}</em>]</span>`).join("<br />")}</details>`
```

Comment:
IIRC, back during the UNO stuff, I was told to split some super long template strings into several smaller chunks and then "concatenate" them